Assignment 2

In the previous assignment, you wrote a program to up-scale an image. In this assign- ments, you will evaluate qualitatively and quantitatively your up-scaling algorithm.

Task: Evaluate your Up-scaling algorithm

(a)  In ILIAS you will find a folder with images at the original size from which you got a down-scale version. Do you observe difference between the two versions? If yes, comment them showing some zooming view.

When we use the code to increase the size of an image, the resulting enlarged image will have twice the size in width and height compared to the original image.

Visually, this can mean that the enlarged image will be sharper and more detailed, but it depends on the quality of the original image and the specific method used to increase the size of the image. However, the quality of the enlarged image may be slightly lower when using the PIL library compared to more complex machine learning or computer vision methods.

(b) To evaluate the quality of your algorithm, we ask you fist to write your own algorithm to subtract two images. And the, applied it on the original image and your own up- scaling image version. Provide an output image. What do you observe?

In [10]:
import os
from PIL import Image

def subtract_images(original_image_path, upscaled_image_path):
    # Open the original image
    original_image = Image.open(original_image_path)
    # Open the upscaled image
    upscaled_image = Image.open(upscaled_image_path)
    # Subtract the original image from the upscaled image
    diff_image = ImageChops.difference(upscaled_image, original_image)
    # Return the difference image
    return diff_image

def subtract_all_images_in_folder(folder_path):
    # Get the list of files in the folder
    file_list = os.listdir(folder_path)
    # Loop through all files in the folder
    for file_name in file_list:
        # Check if the file is an original image file
        if file_name.startswith("img"):
            # Get the full path to the original image
            original_image_path = os.path.join(folder_path, file_name)
            # Create the corresponding upscaled image file name
            upscaled_file_name = "upscaled_" + file_name
            # Get the full path to the upscaled image
            upscaled_image_path = os.path.join(folder_path, upscaled_file_name)
            # Check if the upscaled image file exists
            if upscaled_file_name in file_list:
                # Subtract the original image from the upscaled image
                diff_image = subtract_images(original_image_path, upscaled_image_path)
                # Create a new filename for the difference image
                diff_file_name = "diff_" + file_name
                # Create the full path for the new file
                diff_file_path = os.path.join(folder_path, diff_file_name)
                # Save the difference image to the new file
                diff_image.save(diff_file_path)

# Get the current directory
current_dir = os.getcwd()
# Specify the path to the images folder
images_path = os.path.join(current_dir, "images")

# Subtract original images from their corresponding upscaled images
subtract_all_images_in_folder(images_path)

Provide an output image. What do you observe?

The result of subtracting the original images from their upscaled versions allows for evaluating the quality of the image up-scaling algorithm.

In my case, the resulting image has mostly black pixels, indicating that the up-scaling algorithm created an image with insufficient pixel density. Additionally, there is a lot of noise in the resulting image, suggesting that the algorithm adds unnecessary details to the image.

The goal is to improve the algorithm by reducing the amount of noise and more clearly defining the edges of objects in the resulting image.

(c) To evaluate quantitatively your algorithm, use the Signal Noise Ratio. Write a function to calculate the Signal Noise Ratio (SNR), using to this formula below. 

SNR = 10log10(Psignal/Pnoise)

In [56]:
from PIL import Image
import numpy as np

def read_image(image_path):
    image = Image.open(image_path)
    return image

def calculate_snr(original_image, processed_image):
    original_array = np.array(original_image)
    processed_array = np.array(processed_image)
    
    psignal = np.mean(original_array)
    pnoise = np.sqrt(np.mean((original_array - processed_array)**2))
    
    snr = 10 * np.log10(psignal / pnoise)
    return snr

# Read original and processed images
image_files = [
    ('img1.jpg', 'upscaled_img1.jpg'),
    ('img2.jpg', 'upscaled_img2.jpg'),
    ('img3.jpg', 'upscaled_img3.jpg')
]

for original_file, processed_file in image_files:
    # Read original and processed images
    original_image = read_image(f'images/{original_file}')
    processed_image = read_image(f'images/{processed_file}')

    # Resize processed image to match the size of the original image
    processed_image = processed_image.resize(original_image.size, Image.ANTIALIAS)

    # Calculate SNR
    snr = calculate_snr(original_image, processed_image)

    print(f'{original_file} vs {processed_file}: SNR = {snr:.2f}')


img1.jpg vs upscaled_img1.jpg: SNR = 14.26
img2.jpg vs upscaled_img2.jpg: SNR = 19.26
img3.jpg vs upscaled_img3.jpg: SNR = 15.33


Comment the result.

The higher the SNR value, the better the quality of the restored image. Usually, if the SNR value is above 20-30 dB, the quality of the restored image is considered good.

The results show that the best quality of the restored image with a high SNR value of 19.26 dB was achieved for img2.jpg. For img1.jpg and img3.jpg, the SNR values are lower, indicating lower quality of restoration. In this case, the quality can be improved.